src/Persistence/Mapping/Driver/FileClassLocator.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Persistence\Mapping\Driver;
+
+use AppendIterator;
+use CallbackFilterIterator;
+use Doctrine\Persistence\Mapping\MappingException;
+use FilesystemIterator;
+use InvalidArgumentException;
+use Iterator;
+use RecursiveDirectoryIterator;
+use RecursiveIteratorIterator;
+use ReflectionClass;
+use RegexIterator;
+use SplFileInfo;
+
+use function array_key_exists;
+use function array_map;
+use function assert;
+use function get_debug_type;
+use function get_declared_classes;
+use function is_dir;
+use function preg_quote;
+use function realpath;
+use function sprintf;
+use function str_replace;
+use function str_starts_with;
+
+/**
+ * ClassLocator implementation that uses a list of file names to locate PHP files
+ * and extract class names from them.
+ *
+ * It is compatible with the Symfony Finder component, but does not require it.
+ */
+final class FileClassLocator implements ClassLocator
+{
+    /** @param iterable<SplFileInfo> $files An iterable of files to include. */
+    public function __construct(
+        private iterable $files,
+    ) {
+    }
+
+    /** @return list<class-string> */
+    public function getClassNames(): array
+    {
+        $includedFiles = [];
+
+        foreach ($this->files as $file) {
+            // @phpstan-ignore function.alreadyNarrowedType, instanceof.alwaysTrue
+            assert($file instanceof SplFileInfo, new InvalidArgumentException(sprintf('Expected an iterable of SplFileInfo, got %s', get_debug_type($file))));
+
+            // Skip non-files
+            if (! $file->isFile()) {
+                continue;
+            }
+
+            // realpath() can return false if the file is in a phar archive
+            // @phpstan-ignore ternary.shortNotAllowed
```

Comment:
I have no idea if this is actually true, but here is a comment showing why this should be OK.

```suggestion
            // @phpstan-ignore ternary.shortNotAllowed (false is the only falsy value getRealPath() may return)
```

Reply:
I think the comment just above is sufficient, don't you?

Reply:
The comment above justifies why we should check for `false`, but not why it's OK to use loose comparison.